operator/controllers/kyma_controller.go

```diff
@@ -39,24 +54,135 @@ type KymaReconciler struct {
+
+	if len(kymaObj.Spec.Components) < 1 {
+		return fmt.Errorf("no component specified for resource %s", namespacedName)
+	}
+
+	for _, component := range kymaObj.Spec.Components {
+		componentName := component.Name + "-name"
+
+		componentBytes, ok := configMap.Data[component.Name]
+		if !ok {
+			return fmt.Errorf("%s component not found for resource %s", component.Name, namespacedName)
+		}
+
+		componentYaml := make(map[string]interface{})
+		if err := yaml.Unmarshal([]byte(componentBytes), &componentYaml); err != nil {
+			return fmt.Errorf("error during config map unmarshal %w", err)
+		}
+
+		gvr := schema.GroupVersionResource{
+			Group:    componentYaml["group"].(string),
+			Resource: componentYaml["resource"].(string),
+			Version:  componentYaml["version"].(string),
+		}
+
+		if exists, err := r.GetUnstructuredResource(ctx, gvr, componentName, req.Namespace); client.IgnoreNotFound(err) != nil {
+			return err
+		} else if exists {
+			continue
+		}
+
+		componentUnstructured := &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"kind":       componentYaml["kind"].(string),
+				"apiVersion": componentYaml["group"].(string) + "/" + componentYaml["version"].(string),
+				"metadata": map[string]interface{}{
+					"name":      componentName,
+					"namespace": req.Namespace,
+				},
+				"spec": componentYaml["spec"],
+			},
+		}
+
+		// set owner reference
+		if err := controllerutil.SetOwnerReference(&kymaObj, componentUnstructured, r.Scheme); err != nil {
+			return fmt.Errorf("error setting owner reference on component CR of type: %s for resource %s %w", component.Name, namespacedName, err)
+		}
+
+		if err := r.Client.Create(context.Background(), componentUnstructured, &client.CreateOptions{}); err != nil {
+			return fmt.Errorf("error creating custom resource of type %s %w", component.Name, err)
+		}
+
+		logger.Info("successfully created component CR of", "type", component.Name)
+	}
+
+	return nil
+}
+
+func (r *KymaReconciler) GetUnstructuredResource(ctx context.Context, gvr schema.GroupVersionResource, name string, namespace string) (bool, error) {
+	config, err := GetConfig()
+	if err != nil {
+		return false, err
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return false, err
+	}
+
+	if _, err = dynamicClient.Resource(gvr).Namespace(namespace).Get(ctx, name, metav1.GetOptions{}); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *KymaReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&operatorv1alpha1.Kyma{}).
+	r.Builder = ctrl.NewControllerManagedBy(mgr)
+	return r.Builder.For(&operatorv1alpha1.Kyma{}).
+		//Owns(&corev1.ConfigMap{}).
+		Watches(
+			&source.Kind{Type: &corev1.ConfigMap{}},
+			handler.EnqueueRequestsFromMapFunc(r.findObjectsForConfigMap),
+			builder.WithPredicates(predicate.ResourceVersionChangedPredicate{}),
```

Comment:
I think this is gonna happen quite often, so a watch on this predicate is quite expensive, no?

Reply:
It is sure, but it will be restricted by cache selectors, coming up next :)